plugin/pkg/global/resourcereferencemanager/admission.go

```diff
@@ -292,6 +292,41 @@ func (r *ReferenceManager) Admit(ctx context.Context, a admission.Attributes, o
 				})
 			}
 		}
+	case core.Kind("CloudProfile"):
+		cloudProfile, ok := a.GetObject().(*core.CloudProfile)
+		if !ok {
+			return apierrors.NewBadRequest("could not convert resource into CloudProfile object")
+		}
+		if utils.SkipVerification(operation, cloudProfile.ObjectMeta) {
+			return nil
+		}
+		if a.GetOperation() == admission.Update {
+			oldCloudProfile, ok := a.GetOldObject().(*core.CloudProfile)
+			if !ok {
+				return apierrors.NewBadRequest("could not convert old resource into CloudProfile object")
+			}
+
+			// getting versions removed from the CloudProfile
+			removedVersions := helper.GetRemovedVersions(oldCloudProfile.Spec.Kubernetes.Versions, cloudProfile.Spec.Kubernetes.Versions)
+			if len(removedVersions) > 0 {
+				shootList, err := r.shootLister.List(labels.Everything())
+				if err != nil {
+					return err
+				}
+				usedVersions := make(map[string][]string, len(shootList))
+				for _, shoot := range shootList {
+					if _, ok := usedVersions[shoot.Spec.Kubernetes.Version]; !ok {
+						usedVersions[shoot.Spec.Kubernetes.Version] = append(usedVersions[shoot.Spec.Kubernetes.Version], fmt.Sprintf("name: '%s', provider: '%s'", shoot.Name, shoot.Spec.Provider.Type))
```

Comment:
```suggestion
						usedVersions[shoot.Spec.Kubernetes.Version] = append(usedVersions[shoot.Spec.Kubernetes.Version], fmt.Sprintf("name: '%s', namespace: '%s'", shoot.Name, shoot.Namespace))
```
Or something similar... Provider is not relevant, I guess

Reply:
I think name, namespace & CloudProfile would be better. This check makes sure that a version cannot be deleted from a CloudProfile if a Shoot is still referencing it.

Reply:
Updated the implementation according to the above statement